### PR TITLE
Add optional support for executable space protections

### PR DESCRIPTION
**Pull request by nmeum**
### Contribution description

[Executable space protections][0] are a protection mechanism for
operating system security which mitigate exploitation of buffer
overflows. The exploit payload, used for exploiting buffer overflows, is
often placed in the buffer itself. As an example, consider the exploit
for `sock_dns` from #10739.

Considering a stack-based buffer overflow, the attacker can simply
overwrite the function return address with the address of the buffer
thereby causing the code contained in it to be executed. I've worked on
techniques for preventing the overwrite of the function return address
(#13119, #13175) but these can be circumvented. As such, conventional
operating system also mark the text and data segment non-executable to
prevent attackers from placing exploit payloads there (executable space
protections). Enforcing this efficiently requires a memory protection or
memory management unit.

I noticed recently that RIOT already supports the Cortex-M4 MPU. Using
this MPU on the `nucleo-f401re` I came up with an implementation of
executable space protections for RIOT. The implementation is entirely
optional and must be activated explicitly by using the `mpu_noexec_ram`
pseudomodule. When activated, the entire SRAM section is marked as non
executable, reads- and writes are still allowed.

Kudos to @pyropeter for helping out with this.

### Testing procedure

The changes proposed here also include a test for this pseudo module.
The test is further described in `tests/mpu_noexec_ram/README.md`. The
test successfully passed with `BOARD=nucleo-f401re`.

### Issues/PRs references

Executable space protections can be circumvented through code-reuse
attacks (e.g. [return-to-libc][1] or [ROP][2] in the general case). These attacks
can be mitigated through address randomization techniques such as ASLR.
While ASLR is not deemed employable on constrained devices, a different
address randomization technique, often referred to as link time reordering is.

This technique has been implemented by myself for RIOT in #13176. I
would highly recommended that this is merged as well though this PR does
not technically depend on it.

[0]: https://en.wikipedia.org/wiki/Executable_space_protection
[1]: https://en.wikipedia.org/wiki/Return-to-libc_attack
[2]: https://en.wikipedia.org/wiki/Return-oriented_programming

### CODE REVIEW

**Comment by nmeum** [comment]
Opps, didn't intended to request a review from so many people, sorry.

**Comment by benpicco** [comment]
> Opps, didn't intended to request a review from so many people, sorry.

Not your fault, the [CODEOWNERS](https://github.com/RIOT-OS/RIOT/blob/master/CODEOWNERS) file is responsible for that.
If your PR touches a file or directory with an 'owner' attached, they will be notified.
Currently we have many owners for the entire `tests/` directory - we'll see if this is useful or not (by seeing if those owners submit PRs for more fine grained ownership).

File: tests/mpu_noexec_ram/Makefile [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+BOARD_WHITELIST += samd21-xpro      # cortex-m0plus
+BOARD_WHITELIST += saml21-xpro      # cortex-m0plus
+BOARD_WHITELIST += samr21-xpro      # cortex-m0plus
+BOARD_WHITELIST += sltb001a         # cortex-m4f
+BOARD_WHITELIST += slwstk6220a      # cortex-m4f
+BOARD_WHITELIST += sodaq-autonomo   # cortex-m0plus
+BOARD_WHITELIST += spark-core       # cortex-m3
+BOARD_WHITELIST += stm32f3discovery # cortex-m4f
+BOARD_WHITELIST += stm32f4discovery # cortex-m4f
+BOARD_WHITELIST += udoo             # cortex-m3
```
Comment (benpicco) [inline tests/mpu_noexec_ram/Makefile]:
```suggestion
FEATURES_REQUIRED += arch_cortexm
```
should work as well.

File: tests/mpu_noexec_ram/Makefile.ci [hunk 2]
```diff
@@ -0,0 +1,3 @@
+BOARD_INSUFFICIENT_MEMORY := \
+    stm32f030f4-demo \
+    #
```
Comment (benpicco) [inline tests/mpu_noexec_ram/Makefile.ci]: Not needed - even though the stm32f030f4 is tiny, this test still fits.

File: tests/mpu_noexec_ram/Makefile (hunk 1 above)
Comment (nmeum) [reply tests/mpu_noexec_ram/Makefile]: I don't think many of cortexm boards do not have an MPU.

File: tests/mpu_noexec_ram/Makefile.ci (hunk 2 above)
Comment (nmeum) [reply tests/mpu_noexec_ram/Makefile.ci]: Ah, good to know I copied the two Makefiles from `mpu_stack_guard`. Thanks for checking this! I just removed the `Makefile.ci`.

File: tests/mpu_noexec_ram/Makefile (hunk 1 above)
Comment (benpicco) [reply tests/mpu_noexec_ram/Makefile]: True, the MPU is optional.
The right course of action then would be to add

```
FEATURES_PROVIDED += cortexm_mpu
```

to `cpu/$arch/Makefile.features`

Comment (nmeum) [reply tests/mpu_noexec_ram/Makefile]: Can we do that separately? I would prefer to get the feature merged first and do that cleanup afterwards.

Comment (kaspar030) [reply tests/mpu_noexec_ram/Makefile]: I had this in a branch, opened a WIP PR here: https://github.com/RIOT-OS/RIOT/pull/13391

**Comment by kaspar030** [comment]
Looks good. Note to myself: need to document the use of MPU regions somewhere.

Can you provide a test script that catches the panic (and a possible "I HAVE NOT CRASHED" which would be reached if the MPU did not work as intended)?

**Comment by nmeum** [comment]
> Can you provide a test script that catches the panic (and a possible "I HAVE NOT CRASHED" which would be reached if the MPU did not work as intended)?

This is technically possible, though the latter will be difficult to do sanely since the code in the buffer on the stack would than need to return control graceful back to the C code, jump to a specific C function or something like that.

Maybe that's an improvement which can later be made in a separate PR. Possibly with also updating the `tests/mpu_stack_guard` accordingly?

**Comment by kaspar030** [comment]
> Maybe that's an improvement which can later be made in a separate PR.

Thing is, we don't merge test applications anymore which don't come with a test script.
It would be fine to wait for the correct (expected) panic and timeout otherwise.

**Comment by nmeum** [comment]
> Thing is, we don't merge test applications anymore which don't come with a test script.

Alright, converted the `mpu_noexec_ram` test to an automated test. The test script checks whether the memory management interrupt handler is invoked correctly after jumping to the stack-based buffer.

**Comment by nmeum** [comment]
> tests/mpu_noexec_ram/main.c:32: warning (memsetValueOutOfRange): The 2nd memset() argument '4294967295' doesn't fit into an 'unsigned char'.

Shouldn't `UINT32_MAX` fit into a `uint32_t`?

**Comment by benpicco** [comment]
`unsigned char` is `uint8_t` though.

**Comment by nmeum** [comment]
> `unsigned char` is `uint8_t` though.

Let's just get rid of the memset it's cleaner without anyhow.

**Comment by nmeum** [comment]
Hm? Why was the label removed, the cppcheck warning has been fixed.

**Comment by benpicco** [comment]
Until #13456 is merged all builds will fail now.

File: tests/mpu_noexec_ram/main.c [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+#define JMPBUF_SIZE 3
+
+int main(void)
+{
+    uint32_t buf[JMPBUF_SIZE];
+
+    /* Fill the buffer with invalid instructions */
+    for (unsigned i = 0; i < JMPBUF_SIZE; i++)
+        buf[i] = UINT32_MAX;
```
Comment (kaspar030) [inline tests/mpu_noexec_ram/main.c]:
```
for (unsigned i = 0; i < JMPBUF_SIZE; i++) {
    buf[i] = UINT32_MAX;
}
```
(add scoping block)

Comment (nmeum) [reply tests/mpu_noexec_ram/main.c]: Done.

**Comment by nmeum** [comment]
Rebased this against #13391. All tests pass on the CI now. Merge?

**Comment by benpicco** [comment]
Since this adds a new test to CI, we should let CI run the test first 😉 
But looks good, if everything is green we should merge this!

**Comment by benpicco** [comment]
Did you test this with / without the `mpu_stack_guard` module?

**Comment by nmeum** [comment]
> Did you test this with / without the `mpu_stack_guard` module?

Yes, I did. See 59676a1f5eb4366a6edfd56a20bedbcfcba2276a

**Pull request merged**